File: backend/app/schemas/runs.py
"""Run-related schemas."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class RunSummaryItem(BaseModel):
    """Single runtime record."""

    id: int
    run_uid: str
    run_type: str
    trigger_mode: str
    keyword_id: int | None = None
    keyword: str
    article_id: int | None = None
    status: str
    current_step: str | None = None
    retry_count: int
    error_message: str | None = None
    detail_json: dict[str, Any] | list[Any] | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    updated_at: datetime | None = None


class RunStepItem(BaseModel):
    """Single runtime step."""

    id: int
    job_run_id: int
    step_code: str
    step_name: str
    attempt_no: int
    status: str
    article_id: int | None = None
    error_message: str | None = None
    detail_json: dict[str, Any] | list[Any] | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    updated_at: datetime | None = None


class RunListPayload(BaseModel):
    """Paginated run list payload."""

    items: list[RunSummaryItem] = Field(default_factory=list)
    total: int
    limit: int
    offset: int
    warning: str | None = None


class RunSummaryPayload(BaseModel):
    """Aggregated run summary."""

    total_runs: int
    running_runs: int
    succeeded_runs: int
    failed_runs: int
    partial_runs: int
    latest_run_at: datetime | None = None
    warning: str | None = None


class RunFailuresPayload(BaseModel):
    """Recent failures payload."""

    items: list[RunSummaryItem] = Field(default_factory=list)
    limit: int
    warning: str | None = None


class RunDetailPayload(BaseModel):
    """Detailed run payload."""

    run: RunSummaryItem | None = None
    steps_total: int = 0
    failed_steps: int = 0
    warning: str | None = None


class RunStepsPayload(BaseModel):
    """Single run steps payload."""

    run_id: int
    items: list[RunStepItem] = Field(default_factory=list)
    warning: str | None = None

File: frontend_v2/src/types/runs.ts
import type { PaginatedPayload } from './api'

export interface RunSummaryItem {
  id: number
  run_uid: string
  run_type: string
  trigger_mode: string
  keyword_id: number | null
  keyword: string
  article_id: number | null
  status: string
  current_step: string | null
  retry_count: number
  error_message: string | null
  detail_json: Record<string, unknown> | unknown[] | null
  started_at: string | null
  finished_at: string | null
  updated_at: string | null
}

export interface RunStepItem {
  id: number
  job_run_id: number
  step_code: string
  step_name: string
  attempt_no: number
  status: string
  article_id: number | null
  error_message: string | null
  detail_json: Record<string, unknown> | unknown[] | null
  started_at: string | null
  finished_at: string | null
  updated_at: string | null
}

export type RunListPayload = PaginatedPayload<RunSummaryItem>

export interface RunSummaryPayload {
  total_runs: number
  running_runs: number
  succeeded_runs: number
  failed_runs: number
  partial_runs: number
  latest_run_at: string | null
  warning: string | null
}

export interface RunFailuresPayload {
  items: RunSummaryItem[]
  limit: number
  warning: string | null
}

export interface RunDetailPayload {
  run: RunSummaryItem | null
  steps_total: number
  failed_steps: number
  warning: string | null
}

export interface RunStepsPayload {
  run_id: number
  items: RunStepItem[]
  warning: string | null
}

export interface RunListFilters {
  status?: string
  trigger_mode?: string
  keyword?: string
  limit?: number
  offset?: number
}
